File: p15.py
#!/usr/bin/python
#
# Problem Euler
#
# http://projecteuler.net/index.php?section=problems&id=15
#
# Problem 15
# 19 April 2002
#
# Starting in the top left corner of a 2×2 grid, there are 6 routes (without
# backtracking) to the bottom right corner.
#
# How many routes are there through a 20×20 grid?
# Answer:
#      137846528820

RR = []
for r in range(21):
        RR.append([0]*21)

def R(r, c):
# r stands for 'rows'
# c stands for 'columns'
        if RR[r][c] > 0:
                r=r  # empty operator
        elif r == 1: # just one row
                RR[r][c] = c+1
        elif c == 1: # just one column
                RR[r][c] = r+1
        else: 
                RR[r][c] = R(r-1, c) + R(r, c-1)
        print "%2d %2d -> %d" % (r, c, RR[r][c])
        return RR[r][c]
# end of R

print R(20, 20)

# end of file

File: p15.clj
;; Problem Euler
;;
;; http://projecteuler.net/index.php?section=problems&id=15
;;
;; Problem 15
;; 19 April 2002
;;
;; Starting in the top left corner of a 2×2 grid, there are 6 routes (without
;; backtracking) to the bottom right corner.
;;
;; How many routes are there through a 20×20 grid?
;; Answer:
;;      137846528820

(defn p15
  ([rows cols]
   (let [grid (atom {})
         grid-value (fn [r c] (if-let [[[_ _] x] (find @grid [r c])] x))
         grid-value! (fn [r c v] (swap! grid assoc [r c] v))]
     (letfn [(R [r c]
               (cond (grid-value r c) :nothing
                     (= r 1) (grid-value! r c (+ c 1))
                     (= c 1) (grid-value! r c (+ r 1))
                     :else (grid-value! r c (+ (R (dec r) c) (R r (dec c)))))
               (grid-value r c))]
       (R rows cols))))
  ([]
   (p15 20 20)))


;; end of file
